tools/spark-reader-v1/src/main/java/com/vesoft/nebula/reader/NebulaDataSource.java

```diff
@@ -0,0 +1,50 @@
+/* Copyright (c) 2019 vesoft inc. All rights reserved.
+ *
+ * This source code is licensed under Apache 2.0 License,
+ * attached with Common Clause Condition 1.0, found in the LICENSES directory.
+ */
+
+package com.vesoft.nebula.reader;
+
+import com.vesoft.nebula.bean.ConnectInfo;
+import com.vesoft.nebula.bean.ScanInfo;
+import org.apache.spark.sql.SQLContext;
+import org.apache.spark.sql.sources.BaseRelation;
+import org.apache.spark.sql.sources.RelationProvider;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import scala.collection.immutable.Map;
+import scala.runtime.AbstractFunction0;
+
+public class NebulaDataSource implements RelationProvider {
+
+    private static final Logger LOGGER = LoggerFactory.getLogger(NebulaDataSource.class);
+
+    private ConnectInfo connectInfo;
+    private ScanInfo scanInfo;
+
+    @Override
+    public BaseRelation createRelation(SQLContext sqlContext, Map<String, String> parameters) {
+        // parse and check parameters
+        connectInfo = new ConnectInfo();
+        connectInfo.setIp(parameters.get("ip").get());
+        connectInfo.setSpaceName(parameters.get("spaceName").get());
+        connectInfo.setStoragePort(Integer.parseInt(parameters.get("storagePort").get()));
+        connectInfo.check();
+        LOGGER.info("connectInfo, {}", connectInfo);
+
+        String partitionNumber = parameters.getOrElse("partitionNumber",
+                new AbstractFunction0<String>() {
+                    @Override
+                    public String apply() {
+                        return String.valueOf(Runtime.getRuntime().availableProcessors());
```

Comment:
So the nebula's partition number is same with `availableProcessors` ? 

What's the meaning of this?

Reply:
the partitionNumber is used in NebulaRDD, it mean spark worker number。it can customize，if not, the default value is availableProcessors

    @Override
    public Partition[] getPartitions() {

        int partitionNumber = scanInfo.getPartitionNumber();

        Partition[] partitions = new Partition[partitionNumber];

        for (int i = 0; i < partitionNumber; i++) {

            Partition partition = new NebulaPartition(i, partitionNumber);

            partitions[i] = partition;

        }

        return partitions;

    }


I customize a Partition( NebulaPritition) by extends org.apache.spark.Partition， the NebulaPartition has getScanParts method, is return every worker's scan parts

    public List<Integer> getScanParts(int totalPart) {
        List<Integer> scanParts = new ArrayList<>();
        int currentPart = index + 1;
        while (currentPart <= totalPart) {
            scanParts.add(currentPart);
            currentPart += totalPartition;
        }
        return scanParts;
    }

Reply:
If so... the `partitionNumber ` is required.